### DISCUSSION TITLE
string.residual() not returning explicitly a string, maybe by design

### DISCUSSION THREAD

**Discussion by gilou** [savonet/liquidsoap#3535] [Ideas]
**Describe the bug**
string.residual does not return a string, because it can return null(), and that seems counter-intuitive

**To Reproduce**
`string.residual(prefix='toto_', 'toto_titi)`

```
$ liquidsoap --interactive
[…]
# log(string.residual(prefix='toto_', 'toto_titi'));;
At line 3, char 4-48:

Error 5: this value has type
  _? (inferred at string.liq, line 80, char 2-27)
but it should be a subtype of
  string
```

So I need to string() the output, and I'd say that's because of the null() in [src/libs/string.liq#L108](https://github.com/savonet/liquidsoap/blob/main/src/libs/string.liq#L108) which is somehow reasonable, but makes that function a bit annoying to use if you ask me ;)

**Expected behavior**
Maybe return "" so it can be a string, like string.sub() does?

**Version details**
- OS: Ubuntu 22
- Version: 2.2

**Install method**
Using assets from github actions

**Comment by toots** [comment]
I looked at it and even did the changes but I'm not sure what's the best approach from a user perspective.

While I understand your point of wanting things to be easy, I think that there's also value in guiding the user to make the right decision through the language. Here are some options:

1. Silently return `""` when the string does not begin with the given prefix.

This has the potential to hide important bugs while making the programmer's life easier. This is my least favorite but the most user-friendly.

2. Raise an exception when the string does not begin with the given prefix.

This is also the most user-friendly but, exceptions should really be used sparsely in liquidsoap because we want scripts to run forever without errors.

3. Return `null(()`, the current implementation.

This is the least user-friendly but, it forces the user to think about what to do when the string does not begin with the given prefix while also making sure that there will not be a bad error raised, potentially crashing the script.

Lastly, it's important to remember that we have a coalesce notation to deal with `null` values, so you can always do:

```liquidsoap
string.residual(...) ?? ""
```

This makes it possible to have a default empty string from the function with very little added code.

For these reasons, I'm in favor of keeping things the way they are. I think the extra coalesce is a fair compromise.

**Reply by vitoyucepi** [reply]
I'd like to see more nulls or exceptions in functions.
You may have noticed that I have been trying to use `null.case`, `null.map` and `??` in functions a lot lately.

**Comment by gilou** [comment]
I'm happy with using the coalesce notation… I just usually don't think about it. Full explicit exception handling for that use-case would be avoided smartly by just rewriting string.residual() for that case ;)

We again have many documentation / examples broken due to all that, I'll try to uncover them as I find them…
